### DISCUSSION TITLE
Chainable Actors Extension

### DISCUSSION THREAD

**Discussion by limbusdev** [libktx/ktx#177] [Q&A] [answered]
Hi,
would it be possible to extend Actors to make them chainable?

What we have now:

```kotlin
var myActor = Group()
myActor.setPosition(4f,5f)
myActor.setSize(5f,5f)
```

I have an extension:

```kotlin
fun Actor.lSetSize(width: Float, height: Float): Actor
{
    this.setSize(width, height)
    return this
}

fun Actor.lSetPosition(x: Float, y: Float, align: Int): Actor
{
    this.setPosition(x, y, align)
    return this
}
```

which changes that to

```kotlin
var myActor = Group()
myActor
    .lSetPosition(5f,4f)
    .lSetSize(5f,5f)
```

Problem is, this extension always returns objects of class Actor, not of the inheriting class. Any ideas?

And would you consider ding something similar in libKTX? Or, alternatively, would it be possible to have more Actor-inheriting classes to be defined at top level like tables?

```kotlin
image {
    drawable=pathToDrawable
    position = Vec2(3f,5f)
    ...
}
```

**Comment by jcornaz** [comment, marked as answer]
>this extension always returns objects of class Actor, not of the inheriting class. Any ideas?

You can use generics

```kotlin
fun <T : Actor> T.lSetSize(width: Float, height: Float): T {
    this.setSize(width, height)
    return this
}
```

However, I would personnally not create such extensions, and use `apply` instead:
```kotlin
var myActor = Group().apply {
  setPosition(5f,4f)
  setSize(5f,4f)
}
```

>Or, alternatively, would it be possible to have more Actor-inheriting classes to be defined at top level like tables?

I may be wrong, but I have the impression that it is already provided by the [ktx-scene2d](https://github.com/libktx/ktx/tree/master/scene2d) module.

**Comment by czyzby** [comment]
I advise against writing custom `this`-returning extension for every setter that you plan on using. This approach comes from Java, which lacks the DSL capabilities of Kotlin. Look into `apply` or `let` from the standard library, as @jcornaz mentioned - this would be the idiomatic Kotlin approach.

> And would you consider doing something similar in libKTX?

[`ktx-scene2d`](https://github.com/libktx/ktx/tree/master/scene2d) and [`ktx-vis`](https://github.com/libktx/ktx/tree/master/vis) provide a very similar API using Kotlin type-safe builders DSL. We actually support 100% of LibGDX API, since we didn't write any setter wrappers, but instead provide similar `this` access as the `apply` would. Think of these as syntax sugar for defining actors. Go through their guides, and don't hesitate to ask if you've got any questions.

Does this solve your problem?

**Comment by limbusdev** [comment]
Yeah, thank you. I'm still new to kotlin and didn't know about apply. Thank you for your help. You can close, if you like :)
